### PR TITLE
5320 - Conditional EMDR support

### PR DESCRIPTION
**Pull request by jamesogrant**
Parent (mobile app) issue: https://github.com/senseobservationsystems/goalie-2-mobile-app/issues/5320
Parent (mobile app) PR: https://github.com/senseobservationsystems/goalie-2-mobile-app/pull/5484

### CODE REVIEW

File: android/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -14,4 +14,6 @@
     <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS"/>
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
+    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>
+    <uses-permission android:name="android.permission.BLUETOOTH"/>
```
Comment (xaviraol) [inline android/src/main/AndroidManifest.xml:18]: Do we need those two? Even though it ended up failing in the app side, the main goal of https://github.com/senseobservationsystems/goalie-2-mobile-app/issues/5398 was to remove the requirement of those permissions, so Android 12 users wouldn't be requested for Bluetooth permissions in case they don't have bluetooth headphones connected when having the call, on? Or was it only due to the other permission `BLUETOOTH_CONNECT`?

File: android/src/main/java/com/twiliorn/library/niceday/NDExtra.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import com.twilio.video.TrackPriority;
+import com.twilio.video.TrackSwitchOffMode;
+import com.twilio.video.VideoBandwidthProfileOptions;
+import com.twilio.video.VideoDimensions;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class NDExtra {
```
Comment (xaviraol) [inline android/src/main/java/com/twiliorn/library/niceday/NDExtra.java:32]: It would be nice if we add a bit of documentation explaining what does this NDExtra brings to the table. Or is that explained or detailed somewhere else? It looks like it's a "helper" so CustomTwilioVideoView.java is more clean -and therefore closer to the upstream version- or are there some other reasons?

File: android/src/main/java/com/twiliorn/library/niceday/NDHelper.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package com.twiliorn.library.niceday;
+
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import com.twilio.video.TrackPriority;
+import com.twilio.video.VideoDimensions;
+
+public class NDHelper {
```
Comment (xaviraol) [inline android/src/main/java/com/twiliorn/library/niceday/NDHelper.java:12]: Similar with this one NDHelper. Explain or detail why is this needed in our end to help understand

File: index.d.ts [hunk 4]
```diff
@@ -159,7 +162,7 @@ declare module "react-native-twilio-video-webrtc" {
     enableAudio?: boolean;
     enableVideo?: boolean;
     encodingParameters?: {
-      enableH264Codec: boolean;
+      enableH264Codec?: boolean;
```
Comment (xaviraol) [inline index.d.ts:165]: I see it set it to `false` in current implementation in [app's branch 5320-conditional-emdr](https://github.com/senseobservationsystems/goalie-2-mobile-app/blob/5320-conditional-emdr/app/twilio/sagas.ts#L180).
Any particular reason why we change it to make it optional here? Should we just remove setting it in app's end?

File: src/TwilioVideo.ios.js [hunk 5]
```diff
@@ -166,8 +170,8 @@ export default class TwilioVideo extends Component {
     this._eventEmitter = new NativeEventEmitter(TWVideoModule)
     
     this.setStereoEnabled = this.setStereoEnabled.bind(this)
-    // We expose this to the JS layer to allow avoiding the whole custom audio device code path via CodePush update if there is a critical bug
-    this.usesCustomAudioDevice = true;
+
+    this.useCustomAudioDevice = !!props.useCustomAudioDevice;
```
Comment (xaviraol) [inline src/TwilioVideo.ios.js:174]: Is the double negation necessary here if the `useCustomAudioDevice` is already a `boolean`?

File: ios/RCTTWVideoModule.m [hunk 6]
```diff
@@ -640,6 +641,13 @@ -(TVIVideoBandwidthProfileOptions*)prepareBandwidthProfile:(NSDictionary *)bandw
   [self clearCameraInstance];
   [self stopLocalAudio];
   [self.room disconnect];
+
+  self.camera = nil;
+  self.localVideoTrack = nil;
+  self.localAudioTrack = nil;
+  self.localDataTrack = nil;
+  self.localParticipant = nil;
+  self.room = nil;
```
Comment (xaviraol) [inline ios/RCTTWVideoModule.m:652]: It feels that we already clean some of these properties when calling the methods on top. 
For instance, `localAudioTrack`, `localDataTrack`, and `room`  are already cleared on `stopLocalAudio`. 
We could also call `stopLocalVideo` which would also clear `localVideoTrack`. Maybe we can remove those?

For my understanding, in the upstream they only call `[self.room disconnect]` and the `[self clearCameraInstance]`. Why do we clean up everything else in our fork? Is it just to be sure we have a clean slate when a new call is started?

File: android/src/main/java/com/twiliorn/library/TwilioRemotePreviewManager.java [hunk 7]
```diff
@@ -70,12 +71,12 @@ protected TwilioRemotePreview createViewInstance(ThemedReactContext reactContext
...
-}
+        return MapBuilder.builder()
+                .put(
+                        ON_FRAME_DIMENSIONS_CHANGED,
+                        MapBuilder.of(
+                                "phasedRegistrationNames",
+                                MapBuilder.of("bubbled", ON_FRAME_DIMENSIONS_CHANGED)))
+                .build();
+    }
+}
```
Comment (xaviraol) [inline android/src/main/java/com/twiliorn/library/TwilioRemotePreviewManager.java]: Should we add the end of file line just to not have this annoying sing in GH? 🙃  There are a couple of instances.

**Review by xaviraol — COMMENTED** [review]
Looks good! A few minor comments and questions after looking at the code. 
Testing in a bit!

File: android/src/main/AndroidManifest.xml (hunk 1 above)
Comment (jamesogrant) [reply android/src/main/AndroidManifest.xml:18]: @jagamypriera Can you respond to this? Are these permissions still included upstream?

Comment (jagamypriera) [reply android/src/main/AndroidManifest.xml:18]: those changes are from upstream. it won't show permission dialog to users

File: android/src/main/java/com/twiliorn/library/niceday/NDExtra.java (hunk 2 above)
Comment (jagamypriera) [reply android/src/main/java/com/twiliorn/library/niceday/NDExtra.java:32]: > It looks like it's a "helper" so CustomTwilioVideoView.java is more clean -and therefore closer to the upstream version- or are there some other reasons?

yes, i moved most of our changes on separate package `niceday`

File: index.d.ts (hunk 4 above)
Comment (jagamypriera) [reply index.d.ts:165]: >Any particular reason why we change it to make it optional here?

In the recent upstream, they've added `null` check, so it will be optional.

>I see it set it to false in current implementation in [app's branch 5320-conditional-emdr](https://github.com/senseobservationsystems/goalie-2-mobile-app/blob/5320-conditional-emdr/app/twilio/sagas.ts#L180).

Since it back to optional, we can remove it.

Comment (jamesogrant) [reply index.d.ts:165]: > I see it set it to false in current implementation in [app's branch 5320-conditional-emdr](https://github.com/senseobservationsystems/goalie-2-mobile-app/blob/5320-conditional-emdr/app/twilio/sagas.ts#L180).

If you take a look again I think you'll see it's already gone.

> Any particular reason why we change it to make it optional here? Should we just remove setting it in app's end?

This is an example of a feature that we added ourselves (in the ND branch), that then got added separately upstream. In the upstream repo, this parameter [is optional](https://github.com/blackuy/react-native-twilio-video-webrtc/blob/aa3b9f8c1a011e1df349aa3bc0b53203bde061f7/index.d.ts#L131), so we've made the same change here.

File: src/TwilioVideo.ios.js (hunk 5 above)
Comment (jamesogrant) [reply src/TwilioVideo.ios.js:174]: Yes. It's an optional property, so it could be undefined. `!!` ensures it always evaluates to true/false.

File: android/src/main/java/com/twiliorn/library/niceday/NDExtra.java (hunk 2 above)
Comment (jamesogrant) [reply android/src/main/java/com/twiliorn/library/niceday/NDExtra.java:28]: @jagamypriera Can you respond to this?

File: android/src/main/java/com/twiliorn/library/niceday/NDHelper.java (hunk 3 above)
Comment (jamesogrant) [reply android/src/main/java/com/twiliorn/library/niceday/NDHelper.java:12]: @jagamypriera Can you respond to this?

File: ios/RCTTWVideoModule.m (hunk 6 above)
Comment (jamesogrant) [reply ios/RCTTWVideoModule.m:652]: This is a failsafe added to ensure that Twilio's "internal media factory" is definitely cleaned up (based on advise [here](https://github.com/twilio/video-quickstart-ios/issues/230#issuecomment-367467752)).

Without these additions, Twilio throws an exception when switching from a regular call to a custom audio call. I guess conditional custom audio isn't a common use case, so this bug hasn't been caught upstream.

You're right, we could just identify the specific object that's not cleaned up. But I think it's safer to nil everything here?

File: android/src/main/java/com/twiliorn/library/TwilioRemotePreviewManager.java (hunk 7 above)
Comment (jamesogrant) [reply android/src/main/java/com/twiliorn/library/TwilioRemotePreviewManager.java]: @jagamypriera

File: index.d.ts (hunk 4 above)
Comment (xaviraol) [reply index.d.ts:165]: Cool thanks guys!

File: ios/RCTTWVideoModule.m (hunk 6 above)
Comment (xaviraol) [reply ios/RCTTWVideoModule.m:652]: Clear, yeps I don't mind cleaning everything just to be sure.
Should we have a separate method for it -similar to android NDExtra- to clearly show that this is something on us?

I like the fact that we use ND as a pre-fix, helping us later identify without checking upstream diffs whats ours and whats theirs. 

I leave it up to you guys since you are the ones working on it! Whatever you prefer.

File: android/src/main/java/com/twiliorn/library/niceday/NDExtra.java (hunk 2 above)
Comment (xaviraol) [reply android/src/main/java/com/twiliorn/library/niceday/NDExtra.java:32]: Can we add a few lines on top of the class, or at least a minor explanation of each one of the methods?

**Review by xaviraol — APPROVED** [review]
LGTM! 
Some minor comments about documentation and such left, but nothing blocking at all from my side. Feel free to proceed!

File: ios/RCTTWVideoModule.m (hunk 6 above)
Comment (jamesogrant) [reply ios/RCTTWVideoModule.m:652]: Cool!

> Should we have a separate method for it -similar to android NDExtra- to clearly show that this is something on us?

I don't think so. I think in this instance these changes should probably be included upstream (maybe we should make a PR?).

But I really like the idea of identifying ND additions. Would a comment like `\\ ND Ensure the internal media factory is cleaned up to avoid an exception when switching to custom audio` do for now?
